service/kvmclusterconfig/v1/key/key_test.go

```diff
@@ -0,0 +1,177 @@
+package key
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/giantswarm/apiextensions/pkg/apis/core/v1alpha1"
+	"github.com/giantswarm/microerror"
+)
+
+func Test_ClusterID(t *testing.T) {
+	testCases := []struct {
+		CustomObject v1alpha1.KVMClusterConfig
+		ExpectedID   string
+	}{
+		{
+			CustomObject: v1alpha1.KVMClusterConfig{},
+			ExpectedID:   "",
+		},
+		{
+			CustomObject: v1alpha1.KVMClusterConfig{
+				Spec: v1alpha1.KVMClusterConfigSpec{
+					Guest: v1alpha1.KVMClusterConfigSpecGuest{
+						ClusterGuestConfig: v1alpha1.ClusterGuestConfig{
+							ID: "cluster-1",
+						},
+					},
+				},
+			},
+			ExpectedID: "cluster-1",
+		},
+		{
+			CustomObject: v1alpha1.KVMClusterConfig{
+				Spec: v1alpha1.KVMClusterConfigSpec{
+					Guest: v1alpha1.KVMClusterConfigSpecGuest{
+						ClusterGuestConfig: v1alpha1.ClusterGuestConfig{
+							ID:   "cluster-123",
+							Name: "First cluster",
+						},
+					},
+				},
+			},
+			ExpectedID: "cluster-123",
+		},
+	}
+
+	for i, tc := range testCases {
+		clusterID := ClusterID(tc.CustomObject)
+		if clusterID != tc.ExpectedID {
+			t.Errorf("TestCase %d: ClusterID %s doesn't match. Expected: %s", (i + 1), clusterID, tc.ExpectedID)
+		}
+	}
+}
+
+func Test_EncryptionKeySecretName(t *testing.T) {
+	testCases := []struct {
+		CustomObject       v1alpha1.KVMClusterConfig
+		ExpectedSecretName string
+	}{
+		{
+			CustomObject:       v1alpha1.KVMClusterConfig{},
+			ExpectedSecretName: "-encryption",
+		},
+		{
+			CustomObject: v1alpha1.KVMClusterConfig{
+				Spec: v1alpha1.KVMClusterConfigSpec{
+					Guest: v1alpha1.KVMClusterConfigSpecGuest{
+						ClusterGuestConfig: v1alpha1.ClusterGuestConfig{
+							ID: "cluster-1",
+						},
+					},
+				},
+			},
+			ExpectedSecretName: "cluster-1-encryption",
+		},
+		{
+			CustomObject: v1alpha1.KVMClusterConfig{
+				Spec: v1alpha1.KVMClusterConfigSpec{
+					Guest: v1alpha1.KVMClusterConfigSpecGuest{
+						ClusterGuestConfig: v1alpha1.ClusterGuestConfig{
+							ID:   "cluster-123",
+							Name: "First cluster",
+						},
+					},
+				},
+			},
+			ExpectedSecretName: "cluster-123-encryption",
+		},
+	}
+
+	for i, tc := range testCases {
+		encryptionKeySecretName := EncryptionKeySecretName(tc.CustomObject)
+		if encryptionKeySecretName != tc.ExpectedSecretName {
+			t.Errorf("TestCase %d: EncryptionKeySecretName %s doesn't match. Expected: %s",
+				(i + 1), encryptionKeySecretName, tc.ExpectedSecretName)
+		}
+	}
+
+}
+
+func Test_ToCustomObject(t *testing.T) {
+	var emptyKVMClusterConfigPtr *v1alpha1.KVMClusterConfig
+
+	testCases := []struct {
+		InputObject          interface{}
+		ExpectedCustomObject v1alpha1.KVMClusterConfig
+		ExpectedError        error
+	}{
+		// Success case - pass empty KVMClusterConfig reference
```

Comment:
It would be nice for the humans to have the case index in the test comment. This helps greatly when debugging issues with the implementation because you do not need to count the items in the list but can go directly to the right block either by reading or searching for the test case (this only works when the comments are consistent though).

Reply:
I was thinking about this when writing, but as I used grouping of success/failure cases, the comment indexing easily falls out of sync. How about adding a `name` field to test case and using that instead of counting?

Reply:
I think once tests are written they do not frequently change anymore. You can just go through them at the end and add numbers. See https://github.com/giantswarm/valuemodifier/blob/master/path/path_test.go for instance. The indexing is still old but you see it kind of makes sense. When you want to use names I fear coming up with names will just be more complicated and exhausting than indexing them by number. Do what you think is right. I index them because this was the best approach I came up with so far.

Reply:
@tuommaki related to this take a look at how we took advantage of description fields https://github.com/giantswarm/aws-operator/blob/master/service/awsconfig/v4/resource/cloudformation/adapter/adapter_test.go#L30 used for subtests names https://github.com/giantswarm/aws-operator/blob/master/service/awsconfig/v4/resource/cloudformation/adapter/adapter_test.go#L114 in aws-operator.

Reply:
I ran those tests some day, and the description wasn't showing up :D

Reply:
I noticed that once tests fail the terminal gets spammed with a lot of failures at the same time with this. Usually the tests stop complaining at the first failure. I also think that when we use such sub test magic there should actually only one thing being tested in a sub. Then you have multiple subs and an according description.

Reply:
You need to try harder :)
```
$ go test -v ./service/awsconfig/v4/resource/cloudformation/adapter/
=== RUN   TestAdapterGuestMain
=== RUN   TestAdapterGuestMain/basic_match
=== RUN   TestAdapterGuestMain/different_region
=== RUN   TestAdapterGuestMain/invalid_region
--- PASS: TestAdapterGuestMain (0.00s)
    --- PASS: TestAdapterGuestMain/basic_match (0.00s)
    --- PASS: TestAdapterGuestMain/different_region (0.00s)
    --- PASS: TestAdapterGuestMain/invalid_region (0.00s)
=== RUN   TestAdapterAutoScalingGroupRegularFields
=== RUN   TestAdapterAutoScalingGroupRegularFields/no_worker_nodes,_return_error
=== RUN   TestAdapterAutoScalingGroupRegularFields/basic_matching,_all_fields_present
=== RUN   TestAdapterAutoScalingGroupRegularFields/7_node_cluster,_batch_size_and_min_instances_are_correct
--- PASS: TestAdapterAutoScalingGroupRegularFields (0.00s)
    --- PASS: TestAdapterAutoScalingGroupRegularFields/no_worker_nodes,_return_error (0.00s)
    --- PASS: TestAdapterAutoScalingGroupRegularFields/basic_matching,_all_fields_present (0.00s)
    --- PASS: TestAdapterAutoScalingGroupRegularFields/7_node_cluster,_batch_size_and_min_instances_are_correct (0.00s)
[.......]
```
The `different_region`, `invalid_region`, `7_node_cluster,_batch_size_and_min_instances_are_correct`, ... are the test case descriptions.

Reply:
> I noticed that once tests fail the terminal gets spammed with a lot of failures at the same time with this. Usually the tests stop complaining at the first failure

That probably has to do with using `t.Errorf` instead of `t.Fatalf`, not with subtests AFAIK.

Reply:
I see. Somehow I couldn't find matches, maybe I didn't try hard enough :)

Reply:
> That probably has to do with using `t.Errorf` instead of `t.Fatalf`, not with subtests AFAIK.

I thought the same and tried it out but the result was the same.

Reply:
The indexing doesn't seem to be completely unambiguous. There's actually `i+1` in `valuemodifier` `path_test.go` as well. I like the `description` approach `aws-operator` tests. There is the burden of coming up with names, but the again, I think the comment would need a short description anyway. I'll go with the description and sub-tests, but I'll create a separate PR for this in order to fix all tests in one go.

Oh, and I'm with @fgimenez here that terminal spamming is most likely from `t.Errorf`.